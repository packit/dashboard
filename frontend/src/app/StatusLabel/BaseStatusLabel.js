import React from "react";
import { Label, Tooltip } from "@patternfly/react-core";

import PropTypes from "prop-types";
import { Link } from "react-router-dom";

/**
 * Represents an internal base class for status labels. Handles rendering and basic
 * properties that are given to almost every status label, e.g. link and tooltip.
 */
class BaseStatusLabel extends React.Component {
    static propTypes = {
        link: PropTypes.string,
        status: PropTypes.string,
    };

    /**
     * Constructs basic status label. Tooltip text defaults to status.
     *
     * @param {*} props Properties of the status label, minimal requirements are:
     *  status.
     */
    constructor(props) {
        super(props);
        this.link = props.link;
        this.isExternalLink = props.link && props.link.startsWith("http");
        this.tooltipText = props.status;
    }

    render() {
        return (
            <Tooltip content={this.tooltipText}>
                <span style={{ padding: "2px" }}>
                    <Label
                        icon={this.icon}
                        color={this.color}
                        render={({ className, content, componentRef }) => {
                            return this.isExternalLink ? (
                                <a
                                    href={this.link}
                                    className={className}
                                    target="_blank"
                                    rel="noreferrer"
                                >
                                    {content}
                                </a>
                            ) : (
                                <Link
                                    to={this.link}
                                    className={className}
                                    innerRef={componentRef}
                                >
                                    {content}
                                </Link>
                            );
                        }}
                    >
                        {this.label}
                        <span className="pf-u-screen-reader">
                            {this.tooltipText}
                        </span>
                    </Label>
                </span>
            </Tooltip>
        );
    }
}

export { BaseStatusLabel };
