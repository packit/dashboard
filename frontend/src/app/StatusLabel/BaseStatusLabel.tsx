// Copyright Contributors to the Packit project.
// SPDX-License-Identifier: MIT

import React from "react";
import { Label, Tooltip } from "@patternfly/react-core";
import { Link } from "react-router-dom";

export interface BaseStatusLabelProps {
  link?: string;
  tooltipText: string;
  icon: React.ReactNode;
  color:
    | "blue"
    | "cyan"
    | "green"
    | "orange"
    | "purple"
    | "red"
    | "grey"
    | "gold"
    | undefined;
  label: React.ReactNode;
}

/**
 * Represents an internal base class for status labels. Handles rendering and basic
 * properties that are given to almost every status label, e.g. link and tooltip.
 */
export const BaseStatusLabel: React.FC<BaseStatusLabelProps> = (props) => {
  return (
    <Tooltip content={props.tooltipText}>
      <span style={{ padding: "2px" }}>
        <Label
          icon={props.icon}
          color={props.color}
          render={({ className, content, componentRef }) => {
            // the `downstream_pr_url` can be undefined if
            // looking at propose downstream job details
            if (!props.link) return <>{content}</>;
            return props.link.startsWith("http") ? (
              <a
                href={props.link}
                className={className}
                target="_blank"
                rel="noreferrer"
              >
                {content}
              </a>
            ) : (
              <Link to={props.link} className={className} ref={componentRef}>
                {content}
              </Link>
            );
          }}
        >
          {props.label}
          <span className="pf-v5-u-screen-reader">{props.tooltipText}</span>
        </Label>
      </span>
    </Tooltip>
  );
};
