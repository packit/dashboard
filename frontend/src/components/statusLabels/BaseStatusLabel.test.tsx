// Copyright Contributors to the Packit project.
// SPDX-License-Identifier: MIT

import { IceCreamIcon } from "@patternfly/react-icons";
import { render } from "@testing-library/react";
import { test } from "vitest";
import { BaseStatusLabel } from "./BaseStatusLabel";

test("default label", async ({ expect }) => {
  const { baseElement } = render(
    <BaseStatusLabel
      color="blue"
      tooltipText="tooltipText"
      icon={<IceCreamIcon />}
      label="Label fsdf"
    />,
  );
  expect(baseElement).toMatchSnapshot("default label");
});
