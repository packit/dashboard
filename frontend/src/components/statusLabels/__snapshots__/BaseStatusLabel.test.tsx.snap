// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`default label > default label 1`] = `
<body>
  <div>
    <div
      style="display: contents;"
    >
      <span
        style="padding: 2px;"
      >
        <span
          class="pf-v5-c-label pf-m-blue"
        >
          <span
            class="pf-v5-c-label__icon"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M368 160h-.94a144 144 0 1 0-286.12 0H80a48 48 0 0 0 0 96h288a48 48 0 0 0 0-96zM195.38 493.69a31.52 31.52 0 0 0 57.24 0L352 288H96z"
              />
            </svg>
          </span>
          <span
            class="pf-v5-c-label__text"
          >
            Label fsdf
            <span
              class="pf-v5-u-screen-reader"
            >
              tooltipText
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
</body>
`;
